In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('train_bm.csv')
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [3]:
df.Outlet_Type.value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64

Since this is not an ordinal feature we can one hot encode this variable

In [4]:
pd.get_dummies(df.Outlet_Type).head()

Grocery Store  Supermarket Type1  Supermarket Type2  Supermarket Type3
0              0                  1                  0                  0
1              0                  0                  1                  0
2              0                  1                  0                  0
3              1                  0                  0                  0
4              0                  1                  0                  0

Applying one hot encoding on Entire Dataset

In [5]:
df_encoded = pd.get_dummies(df)
df_encoded.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
0         9.30         0.016047  249.8092                       1999   
1         5.92         0.019278   48.2692                       2009   
2        17.50         0.016760  141.6180                       1999   
3        19.20         0.000000  182.0950                       1998   
4         8.93         0.000000   53.8614                       1987   

   Item_Outlet_Sales  Item_Identifier_DRA12  Item_Identifier_DRA24  \
0          3735.1380                      0                      0   
1           443.4228                      0                      0   
2          2097.2700                      0                      0   
3           732.3800                      0                      0   
4           994.7052                      0                      0   

   Item_Identifier_DRA59  Item_Identifier_DRB01  Item_Identifier_DRB13  ...  \
0                      0                      0                      0  ...   
1                      0                      0                      0  ...   
2                      0                      0                      0  ...   
3                      0                      0                      0  ...   
4                      0                      0                      0  ...   

   Outlet_Size_High  Outlet_Size_Medium  Outlet_Size_Small  \
0                 0                   1                  0   
1                 0                   1                  0   
2                 0                   1                  0   
3                 0                   0                  0   
4                 1                   0                  0   

   Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
0                            1                            0   
1                            0                            0   
2                            1                            0   
3                            0                            0   
4                            0                            0   

   Outlet_Location_Type_Tier 3  Outlet_Type_Grocery Store  \
0                            0                          0   
1                            1                          0   
2                            0                          0   
3                            1                          1   
4                            1                          0   

   Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
0                              1                              0   
1                              0                              1   
2                              1                              0   
3                              0                              0   
4                              1                              0   

   Outlet_Type_Supermarket Type3  
0                              0  
1                              0  
2                              0  
3                              0  
4                              0  

[5 rows x 1605 columns]

But This creates Two problems

Look at the newly created features --> Outlet_Size_High,	Outlet_Size_Medium,	Outlet_Size_Small

In [6]:
df_encoded[['Outlet_Size_High', 'Outlet_Size_Medium', 'Outlet_Size_Small']].head()

Outlet_Size_High  Outlet_Size_Medium  Outlet_Size_Small
0                 0                   1                  0
1                 0                   1                  0
2                 0                   1                  0
3                 0                   0                  0
4                 1                   0                  0

The feature Outlet_Size is an ordinal Variable. Hence, Doing One Hot encoding on these variables there is a loss
in information or heirarchy.

Problem 2

In [7]:
df.shape, df_encoded.shape

((8523, 12), (8523, 1605))

The number of features has increased from 12 to 1605. This is because of the feature -> Item_Identifier. Due to this there is redundunt number of zeros or values. 

In [8]:
df_encoded[
    ['Item_Identifier_DRA12', 'Item_Identifier_DRA24',
       'Item_Identifier_DRA59', 'Item_Identifier_DRB01',
       'Item_Identifier_DRB13', 'Item_Identifier_DRB24',
       'Item_Identifier_DRB25', 'Item_Identifier_DRB48',
       'Item_Identifier_DRC01', 'Item_Identifier_DRC12',
       'Item_Identifier_DRC13', 'Item_Identifier_DRC24',
       'Item_Identifier_DRC25', 'Item_Identifier_DRC27',
       'Item_Identifier_DRC36', 'Item_Identifier_DRC49',
       'Item_Identifier_DRD01', 'Item_Identifier_DRD12',
       'Item_Identifier_DRD13', 'Item_Identifier_DRD15',
       'Item_Identifier_DRD24', 'Item_Identifier_DRD25',
       'Item_Identifier_DRD27', 'Item_Identifier_DRD37',
       'Item_Identifier_DRD49']
].head()

Item_Identifier_DRA12  Item_Identifier_DRA24  Item_Identifier_DRA59  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRB01  Item_Identifier_DRB13  Item_Identifier_DRB24  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRB25  Item_Identifier_DRB48  Item_Identifier_DRC01  \
0                      0                      0                      0   
1                      0                      0                      1   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRC12  ...  Item_Identifier_DRC49  Item_Identifier_DRD01  \
0                      0  ...                      0                      0   
1                      0  ...                      0                      0   
2                      0  ...                      0                      0   
3                      0  ...                      0                      0   
4                      0  ...                      0                      0   

   Item_Identifier_DRD12  Item_Identifier_DRD13  Item_Identifier_DRD15  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRD24  Item_Identifier_DRD25  Item_Identifier_DRD27  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   Item_Identifier_DRD37  Item_Identifier_DRD49  
0                      0                      0  
1                      0                      0  
2                      0                      0  
3                      0                      0  
4                      0                      0  

[5 rows x 25 columns]

Solutions For the above problems

Problem 1

In [9]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [10]:
df.Outlet_Size.value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [11]:
le.fit_transform(['small', 'medium', 'high'])

array([2, 1, 0], dtype=int64)

Since Label encoder uses alphabetical order to assign values we cannot use LabelEncoder. Instead We use Map function

In [12]:
df.Outlet_Size = df.Outlet_Size.map( { 'Small' : 0, 'Medium': 1, 'High': 2 } )

In [13]:
df.Outlet_Size.head()

0    1.0
1    1.0
2    1.0
3    NaN
4    2.0
Name: Outlet_Size, dtype: float64

Problem 2

In [14]:
df.Item_Identifier.value_counts()

FDW13    10
FDG33    10
NCB18     9
FDG09     9
NCQ06     9
         ..
FDE52     1
FDC23     1
FDO33     1
FDQ60     1
FDK57     1
Name: Item_Identifier, Length: 1559, dtype: int64

We can see that there are Items with identifiers less than 2.

In [15]:
Item_Identifier_values = df.Item_Identifier
Item_Identifier_values

0       FDA15
1       DRC01
2       FDN15
3       FDX07
4       NCD19
        ...  
8518    FDF22
8519    FDS36
8520    NCJ29
8521    FDN46
8522    DRG01
Name: Item_Identifier, Length: 8523, dtype: object

In [16]:
counts = df.Item_Identifier.value_counts()
counts

FDW13    10
FDG33    10
NCB18     9
FDG09     9
NCQ06     9
         ..
FDE52     1
FDC23     1
FDO33     1
FDQ60     1
FDK57     1
Name: Item_Identifier, Length: 1559, dtype: int64

In [17]:
for index, item in enumerate(Item_Identifier_values):
    if counts[item] < 4:
        Item_Identifier_values[index] = "other"

<ipython-input-17-fe1c6e9c1572>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Item_Identifier_values[index] = "other"


In [18]:
Item_Identifier_values.value_counts()

other    418
FDW13     10
FDG33     10
FDO19      9
FDG09      9
        ... 
FDF40      4
FDK50      4
NCW18      4
NCG54      4
FDS50      4
Name: Item_Identifier, Length: 1403, dtype: int64

In [19]:
df.Item_Identifier = Item_Identifier_values
df.head(7)

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15        9.300          Low Fat         0.016047   
1           DRC01        5.920          Regular         0.019278   
2           FDN15       17.500          Low Fat         0.016760   
3           FDX07       19.200          Regular         0.000000   
4           NCD19        8.930          Low Fat         0.000000   
5           other       10.395          Regular         0.000000   
6           FDO10       13.650          Regular         0.012741   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   
5           Baking Goods   51.4008            OUT018   
6            Snack Foods   57.6588            OUT013   

   Outlet_Establishment_Year  Outlet_Size Outlet_Location_Type  \
0                       1999          1.0               Tier 1   
1                       2009          1.0               Tier 3   
2                       1999          1.0               Tier 1   
3                       1998          NaN               Tier 3   
4                       1987          2.0               Tier 3   
5                       2009          1.0               Tier 3   
6                       1987          2.0               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052  
5  Supermarket Type2           556.6088  
6  Supermarket Type1           343.5528